torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -442,10 +484,10 @@ ProcessGroupNCCL::WorkNCCL::WorkNCCL(
   // Note: The actual events are lazily created when first recorded to with
   // DEFAULT_FLAGS = cudaEventDisableTiming.
   if (enableTiming) {
-    ncclStartEvent_ = std::make_shared<at::cuda::CUDAEvent>(cudaEventDefault);
+    ncclStartEvent_ =
+        CUDAEventCache::get().create(device.index(), enableTiming);
   }
-  ncclEndEvent_ = std::make_shared<at::cuda::CUDAEvent>(
-      enableTiming ? cudaEventDefault : cudaEventDisableTiming);
+  ncclEndEvent_ = CUDAEventCache::get().create(device.index(), enableTiming);
```

Comment:
Any idea how big a performance difference there will be between disableTiming and enableTiming?

Reply:
Here is one analysis I've seen:

https://github.com/harrism/cuda_event_benchmark

Using the items_per_second column, recording an event with timing is 2.5us but without timing it is .25us. Similarly it is about ~10x to cache the events vs cudaEventCreate/Destroy each time.